doc/proposals/tls-utilities.md

```diff
@@ -0,0 +1,159 @@
+
+	Outputs:
+
+        * A ConfigMap that contains the CA Cert:
+
+	```
+	kind: ConfigMap
+	apiVersion: v1
+	metadata:
+  	name: <crd-kind>-<crd-name>-ca
+  	namespace: <ns>
+	data:
+  	ca.crt: ...
+            ```
+        * A secret that contains the CA key:
+
+           ```
+ 	kind: Secret
+ 	apiVersion: v1
+ 	metadata:
+  	name: <cr-kind>-<cr-name>-ca
+  	namespace: <ns>
+ 	data:
+  	ca.key: ...
+	```
+
+2. Verify the CA certificate: The newly generated CA certificate in ConfigMap should be validated i.e. the digital signature should be checked, the expiry, activation dates and validity period should be checked, etc.
+
+3. Generate Server or Client Certificate: We assume that the components inside an application communicate through a service. Therefore, we generate a TLS private key and certificate for that service.
+
+	Inputs:
+	
+	* CRD-kind
+	* CRD-name
+	* Namespace
+	* Previously generated CA key and Certificate
+
+	Options:
+
+	* Key usage: [Server, Client or Both] If Server, needs svc obj (cluster.local can be different). If client, needs CN and possible org.
+	* Option to specify signature configuration policy i.e Signing Profile for the certificate.
+
+        Outputs:
+
+	* A Secret containing Private Key and Server/Client Certificate signed by the CA:
+
+	```
+	kind: Secret
+	apiVersion: v1
+	metadata:
+	 name: <cr-kind>-<cr-name>-<CertName>
+	 namespace: <ns>
+	data:
+	 tls.crt: ...
+	 tls.key: …
+	type: kubernetes.io/tls
+	```
+
+4. Verify Server or Client Certificate: The newly generated client/server certificate in should be validated i.e. the digital signature should be checked, the expiry, activation dates and validity period should be checked, certificate chain should be verified etc
+
+
+5. Use CA configmap and TLS secret in the necessary deployment/pod object.
+
+## Implementation:
+
+Have a certificate configuration object which will contain all the input arguments required to generate a server/client certificate.
+
+```
+// CertConfig configures the Cert generation.
+type CertConfig struct {
+	// CertName is the name of the cert.
+	CertName string
+	// Optional CertType. Server, client or both; defaults to both.
+	CertType CertType
+	// CommonName is the common name of the cert
+	CommonName string
+	// Organization is Organization of the cert
+	Organization []string
+	// Optional CA Key, if user wants to provide custom CA
+	CAKey string
+	// Optional CA Certificate, if user wants to provide custom CA
+	CACert
+}
+```
+
+Have a method which will generate a secret that contains a newly created TLS Private key and Certificate for the given service. A unique CA is also generated by default to sign the Certificate if a prior CA certificate is not preset for the user specified CRD. If the user wants to use a custom CA, it can be specified in the `CertConfig` object. The function signature of the method is as follows:
+
+```
+func GenerateCert(crd runtime.Object, service v1.Service, config CertConfig) (v1.Secret, error)
```

Comment:
@rithujohn191 Can you update the function signature according to the revisions made in #374 
https://github.com/operator-framework/operator-sdk/blob/master/pkg/tlsutil/tls.go#L112
We now return the CA secret(key) and configmap(cert) as well.

You might also want to double check the workflow section or any other comments that might need to be updated in this doc according to  #374